File: core/src/validators/rdfa.ts
// core/src/validators/rdfa.ts
//
// RDFa validator orchestrator. Same shape as microdata.ts: runs the
// extractor, delegates per-item validation to per-item.ts, aggregates
// issues, applies multi-item path-prefix disambiguation, returns one
// ValidationResult with `format: "rdfa"`.

import type {
  Issue,
  ValidationResult,
  ValidateOptions,
} from "../types.js";
import { loadRegistry } from "../registry.js";
import { loadCuratedRules } from "../curated-rules.js";
import { extractRdfa } from "../utils/rdfa-extractor.js";
import { validateItem } from "./per-item.js";

const registry = loadRegistry();
const curatedRules = loadCuratedRules();

function bucket(issues: Issue[]): {
  errors: Issue[];
  warnings: Issue[];
  info: Issue[];
} {
  const errors: Issue[] = [];
  const warnings: Issue[] = [];
  const info: Issue[] = [];
  for (const issue of issues) {
    if (issue.type === "error") errors.push(issue);
    else if (issue.type === "warning") warnings.push(issue);
    else info.push(issue);
  }
  return { errors, warnings, info };
}

function registryProvenance() {
  return {
    schemaVersion: registry.schemaVersion,
    snapshotAt: registry.snapshotAt,
    curatedRulesVersion: curatedRules.sourceVersion,
  };
}

function prefixPath(issue: Issue, typeName: string, prefix: string): Issue {
  let newPath = issue.path;
  if (issue.path === typeName) {
    newPath = prefix;
  } else if (issue.path.startsWith(`${typeName}.`)) {
    newPath = prefix + issue.path.slice(typeName.length);
  }
  if (newPath === issue.path) return issue;
  return { ...issue, path: newPath };
}

export function validateRdfa(
  input: string,
  options: ValidateOptions = {},
): ValidationResult {
  const { items, extractionIssues } = extractRdfa(input);

  const typeCounts = new Map<string, number>();
  for (const item of items) {
    const t = item["@type"];
    if (typeof t === "string") {
      typeCounts.set(t, (typeCounts.get(t) ?? 0) + 1);
    }
  }

  const allIssues: Issue[] = [...extractionIssues];
  const types: string[] = [];
  const typeIndex = new Map<string, number>();

  for (const item of items) {
    const { issues, resolvedType } = validateItem(
      item as Record<string, unknown>,
    );

    if (resolvedType !== null && (typeCounts.get(resolvedType) ?? 0) > 1) {
      const idx = typeIndex.get(resolvedType) ?? 0;
      typeIndex.set(resolvedType, idx + 1);
      const prefix = `${resolvedType}[${idx}]`;
      for (const issue of issues) {
        allIssues.push(prefixPath(issue, resolvedType, prefix));
      }
    } else {
      allIssues.push(...issues);
    }

    if (resolvedType !== null) types.push(resolvedType);
  }

  const { errors, warnings, info } = bucket(allIssues);
  const strictFail = options.strict === true && warnings.length > 0;

  return {
    valid: errors.length === 0 && !strictFail,
    format: "rdfa",
    types,
    errors,
    warnings,
    info,
    registry: registryProvenance(),
  };
}

File: python/src/schema_audit/validators/rdfa.py
"""RDFa orchestrator.

Ports ``core/src/validators/rdfa.ts``. Same per-item engine and same
multi-item path-prefix logic as the Microdata orchestrator.
"""

from __future__ import annotations

from .._per_item import validate_item
from .._rdfa_extractor import extract_rdfa
from ..curated_rules import load_curated_rules
from ..registry import load_registry
from ..types import Issue, RegistryInfo, ValidationResult

_REGISTRY = load_registry()
_CURATED = load_curated_rules()


def _registry_provenance() -> RegistryInfo:
    return {
        "schemaVersion": _REGISTRY["schemaVersion"],
        "snapshotAt": _REGISTRY["snapshotAt"],
        "curatedRulesVersion": _CURATED["sourceVersion"],
    }


def _bucket(issues: list[Issue]) -> tuple[list[Issue], list[Issue], list[Issue]]:
    errors: list[Issue] = []
    warnings: list[Issue] = []
    info: list[Issue] = []
    for issue in issues:
        severity = issue["type"]
        if severity == "error":
            errors.append(issue)
        elif severity == "warning":
            warnings.append(issue)
        else:
            info.append(issue)
    return errors, warnings, info


def _prefix_path(issue: Issue, type_name: str, prefix: str) -> Issue:
    path = issue["path"]
    if path == type_name:
        new_path = prefix
    elif path.startswith(f"{type_name}."):
        new_path = prefix + path[len(type_name) :]
    else:
        return issue
    return {
        "type": issue["type"],
        "code": issue["code"],
        "path": new_path,
        "message": issue["message"],
        "value": issue["value"],
    }


def validate_rdfa(html: str, *, strict: bool = False) -> ValidationResult:
    items, extraction_issues = extract_rdfa(html)

    type_counts: dict[str, int] = {}
    for item in items:
        t = item.get("@type")
        if isinstance(t, str):
            type_counts[t] = type_counts.get(t, 0) + 1

    all_issues: list[Issue] = list(extraction_issues)
    types_out: list[str] = []
    type_index: dict[str, int] = {}

    for item in items:
        result = validate_item(item)

        if result.resolved_type is not None and type_counts.get(result.resolved_type, 0) > 1:
            idx = type_index.get(result.resolved_type, 0)
            type_index[result.resolved_type] = idx + 1
            prefix = f"{result.resolved_type}[{idx}]"
            for issue in result.issues:
                all_issues.append(_prefix_path(issue, result.resolved_type, prefix))
        else:
            all_issues.extend(result.issues)

        if result.resolved_type is not None:
            types_out.append(result.resolved_type)

    errors, warnings, info = _bucket(all_issues)
    strict_fail = strict and len(warnings) > 0

    return {
        "valid": len(errors) == 0 and not strict_fail,
        "format": "rdfa",
        "types": types_out,
        "errors": errors,
        "warnings": warnings,
        "info": info,
        "registry": _registry_provenance(),
    }
